common/defs.go

```diff
@@ -82,8 +108,261 @@ type PortUpdateMessage struct {
 	Port uint64 `json:"port"`
 }
 
+// Endpoint represents an endpoint - that is, something that
+// has an IP address and routes to/from. It can be a container,
+// a Kubernetes POD, a VM, etc.
+type Endpoint struct {
+	Any     string `json:"any,omitempty"`
+	Peer    string `json:"peer,omitempty"`
+	CidrStr string `json:"cidr,omitempty"`
+	// TODO this can be collapsed into Cidr but needs
+	// work on JSON marshaller/unmarshaller to do that.
+	Cidr              net.IPNet `json:"-"`
+	TenantID          uint64    `json:"tenant_id,omitempty"`
+	TenantExternalID  string    `json:"tenant_external_id,omitempty"`
+	TenantNetworkID   *uint64   `json:"tenant_network_id,omitempty"`
+	SegmentID         uint64    `json:"segment_id,omitempty"`
+	SegmentExternalID string    `json:"segment_external_id,omitempty"`
+	SegmentNetworkID  *uint64   `json:"segment_network_id,omitempty"`
+}
+
+func (e Endpoint) String() string {
+	return String(e)
+}
+
+const (
+	PolicyDirectionIngress = "ingress"
+	PolicyDirectionEgress  = "egress"
+)
+
+type PortRange [2]uint
+
+func (p PortRange) String() string {
+	return fmt.Sprintf("%d-%d", p[0], p[1])
+}
+
+// Rule describes a rule of the policy. The following requirements apply
+// (the policy would not be validated otherwise):
+// 1. Protocol must be specified.
+// 2. Protocol must be one of those validated by isValidProto().
+// 3. Ports cannot be negative or greater than 65535.
+// 4. If Protocol specified is "icmp", Ports and PortRanges fields should be blank.
+// 5. If Protocol specified is not "icmp", Icmptype and IcmpCode should be unspecified.
+type Rule struct {
+	Protocol   string      `json:"protocol"`
+	Ports      []uint      `json:"ports,omitempty"`
+	PortRanges []PortRange `json:"port_ranges,omitempty"`
+	// IcmpType only applies if Protocol value is ICMP and
+	// is mutually exclusive with Ports or PortRanges
+	IcmpType   uint `json:"icmp_type,omitempty"`
+	IcmpCode   uint `json:"icmp_code,omitempty"`
+	IsStateful bool `json:"is_stateful,omitempty"`
+}
+
+func (r Rule) String() string {
+	return String(r)
+}
+
+type Rules []Rule
+
+// Metadata attached to entities for various external environments like Open Stack / Kubernetes
+type Tag struct {
+	Key   string `json:"key,omitempty"`
+	Value string `json:"value,omitempty"`
+}
+
+// Policy describes Romana network security policy.
+// For examples, see:
+// 1. https://github.com/romana/core/blob/master/policy/policy.sample.json
```

Comment:
These two examples show different versions, rules with different syntax. Are we sure that both of those are valid?